# Problem 18

https://projecteuler.net/problem=18

In [3]:
rawPyramid = '''75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23'''
# the maximum depth and breadth is 15

We try a DP (dynammic programming) approach again. The idea is that we can break this problem down into multiple smaller sub-arrays/sub-problems, and rather than computing every single path, just look at the maximal pyramid. 

In [ ]:
rawArray =  [
    [int(x) for x in row.split()]
    for row in rawPyramid.splitlines()
]
# we have transformed the raw data from a 
# multi-line string to a proper array

In [8]:
dpArray = [[0] * len(row) for row in rawArray]
dpArray[-1] = rawArray[-1][:]
# this is where we actually store our DP results. 
# idea is we compute every possible sub-pyramid

Now that we have our initial/starting data, we may proceed!

In [12]:
for i in range(len(rawArray) - 2, -1, -1):
    for j in range(len(rawArray[i])):
        dpArray[i][j] = rawArray[i][j] + max(dpArray[i+1][j], dpArray[i+1][j+1])
print(dpArray)

[[1074], [995, 999], [818, 900, 935], [704, 801, 853, 792], [686, 640, 766, 731, 782], [666, 614, 636, 684, 660, 717], [647, 501, 613, 609, 533, 657, 683], [559, 499, 479, 536, 514, 526, 594, 616], [460, 434, 419, 475, 508, 470, 510, 524, 487], [419, 365, 393, 387, 419, 425, 430, 376, 454, 322], [378, 317, 231, 321, 354, 372, 393, 354, 360, 293, 247], [325, 246, 187, 178, 256, 329, 273, 302, 263, 242, 193, 233], [255, 235, 154, 150, 140, 179, 256, 209, 224, 172, 174, 176, 148], [125, 164, 102, 95, 112, 123, 165, 128, 166, 109, 122, 147, 100, 54], [4, 62, 98, 27, 23, 9, 70, 98, 73, 93, 38, 53, 60, 4, 23]]


We see our final solution is 1074. 

TODO: reproduce array with red markers for maximal elements like in the Project Euler problem statement

## Extension
Let's try Problem 67 immediately (I can hear the jokes...)